### PR TITLE
[RDY] Add option to run best partially-matching command

### LINKED ISSUE #623: Auto-use the first match for a command
Requested by @Earnestly in IRC:

Typing, e.g. `:do`<kbd>Enter</kbd> should invoke `:download` automatically as that's the first-best match instead of showing an invalid command error.

(of course configurable)

### PR DESCRIPTION
**Pull request by 7lb**
This fixes #623

Previously, trying to run a non-existent but partially-matching command would fail with "no such command" error, unless the number of partial matches was exactly one.

This adds the `completions.use_best_match` config (defaults to `false`) which if active uses the first of the completions suggestions even if more than one exists.

To illustrate, with `completions.use_best_match` set to `false` (default behaviour), typing `:do<enter>` would result in `no such command`

With `completions.use_best_match` set to `true` (new behaviour), typing `:do<enter>` results in running the `:download` command, since that's the best match

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3063)

### CODE REVIEW

**Comment by 7lb** [comment]
I marked this as WIP because I still did not write tests, but I'm not sure how to proceed in that regard, I'd apprecieate a little push in the right direction, like if there are any similar tests I can look at to produce additional ones, and where to place them.

Also, running `tox` on the master branch produces 107 failures on my machine, which is the same result I get running tests on this branch thus there should be no regressions.

On that note, is it expected to have failing tests?

File: qutebrowser/config/configdata.yml [hunk 1]
```diff
@@ -656,6 +656,11 @@ completion.web_history_max_items:
 
     0: no history / -1: unlimited
 
+completion.use_best_match:
+  type: Bool
+  default: true
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.yml]: I think it'd be better to default this to `false`, to keep the existing behavior.

**Comment by The-Compiler** [comment]
Take a look at `tests/unit/commands/test_runners.py` - `test_partial_parsing` is already quite similar there.

It's not expected to have failing tests, though sometimes a test or two can be flaky. Do you have a new QtWebKit available? If not, you might want to run with something like `tox -e py36 -- --qute-bdd-webengine` to force using QtWebEngine (see #3051).

**Comment by 7lb** [comment]
Hi, I added a couple tests looking at `tests/unit/commands/test_runners.py` and `tests/unit/config/test_configcommands.py`.

I think the best place to put the test is in the latter of the above files, but by all means correct me if I'm wrong.

`test_use_best_match` fails but I'm not sure why. Running qutebrowser and starting the command `:do` I see the first entry in the completion model is `download`, while the test fails because it expects `cancel-download`.

Is the completion model populated differently in the test environment? And what's the best course of action here?

I could just `assert result.cmd.name == 'cancel-download'` in the test, but that doesn't seem right since it should really be `download`, unless the test env is deterministically different.

**Comment by The-Compiler** [comment]
`test_configcommands.py` is the wrong file - it's for tests for `:config-*`, `:set` and `:bind` commands. `test_runners.py` is more appropriate, as it tests `CommandRunner`, and this change happened there.

It's possible that different commands are registered in the test environment. Just add the tests there for now, and I'll take care of the rest (i.e. making the command list for those tests deterministic).

File: tests/unit/config/test_configcommands.py [hunk 2]
```diff
@@ -532,3 +532,32 @@ def test_unbind_invalid(self, commands, key, mode, expected):
...
+
+    """Tests for completions.use_best_match."""
+
+    def test_dont_use_best_match(self, config_stub, monkeypatch):
+        """Test multiple completion options with use_best_match set to false.
+
+        Should raise NoSuchCommandError
+        """
+        config_stub.val.completion.use_best_match = False
+        monkeypatch.setattr('qutebrowser.config', config_stub)
```
Comment (The-Compiler) [inline tests/unit/config/test_configcommands.py]: `config_stub` already does this patching automagically.

**Comment by 7lb** [comment]
Ok, I moved the tests to `tests/unit/commands/test_runners.py` and made them pass by asserting `result.cmd.name == 'download-cancel'`.

I also marked the PR as ready.

**Comment by The-Compiler** [comment]
Thanks for the follow up! It might take me a bit until I get to reviewing this again, since I was really busy with the v1.0 release and the attention it has gotten over the past few days, so things started to pile up a bit.

**Comment by The-Compiler** [comment]
Thanks, fixed the conflicts and merged! I added the tests in 4f263505eea1d35660bd9180b3fd3aaf599277d0.

**Comment by The-Compiler** [comment]
I thought this was inconsistent because of not all commands being registered in tests, but now realized it was just picking a random match, depending on how the dict was ordered in memory:

```python
matches = []
for valid_command in cmdutils.cmd_dict:
    if valid_command.find(cmdstr) == 0:
        matches.append(valid_command)
```

I cleaned that up to the equivalent:

```python
matches = [cmd for cmd in cmdutils.cmd_dict if cmdstr in cmd]
```

and now did this to ensure it actually uses the shortest one in 12c959052452587a559bb55c717ced5321594474:

```python
matches = [cmd for cmd in sorted(cmdutils.cmd_dict, key=len)
           if cmdstr in cmd]
```

**Comment by 7lb** [comment]
Thank you for the updates! Definitely helps with growing my understanding of the code

**Pull request merged**